xdsl/dialects/gpu.py

```diff
@@ -0,0 +1,124 @@
+from __future__ import annotations
+from typing import Generic, Type, TypeVar
+
+from xdsl.ir import Attribute, Operation, Dialect, ParametrizedAttribute
+from xdsl.irdl import ParameterDef, irdl_op_definition, irdl_attr_definition, SingleBlockRegion, OpAttr
+from xdsl.dialects.builtin import StringAttr, SymbolRefAttr
+from xdsl.parser import BaseParser
+from xdsl.printer import Printer
+from xdsl.utils.exceptions import VerifyException
+
+
+@irdl_attr_definition
+class _AllReduceOperationAttr(ParametrizedAttribute):
+    name = "all_reduce_op"
+
+    param: ParameterDef[StringAttr]
+
+    def print_parameters(self, printer: Printer) -> None:
+        printer.print(f"all_reduce_op {self.param.data}")
+
+
+@irdl_attr_definition
+class _DimensionAttr(ParametrizedAttribute):
+    name = "dim"
+
+    param: ParameterDef[StringAttr]
+
+    def print_parameters(self, printer: Printer) -> None:
+        printer.print(f"dim {self.param.data}")
+
+
+T = TypeVar('T',
+            bound=_AllReduceOperationAttr | _DimensionAttr,
+            covariant=True)
+
+
+@irdl_attr_definition
+class _GPUAttr(ParametrizedAttribute, Generic[T]):
+    name = "gpu"
+
+    value: ParameterDef[T]
+
+    @staticmethod
+    def parse_parameters(parser: BaseParser) -> list[Attribute]:
+        parser.parse_characters("<", f"Expected <")
```

Comment:
I would like to see more context provided in this error message, something like ``"GPU Attribute parameters expected here (`<`)"``

Reply:
Makes sense, thanks, I'll add it